File: direct/test/java/ru/yandex/direct/core/testing/info/CampaignInfo.java
package ru.yandex.direct.core.testing.info;

import ru.yandex.direct.core.testing.steps.campaign.model0.Campaign;
import ru.yandex.direct.dbutil.model.ClientId;

import static ru.yandex.direct.utils.CommonUtils.ifNotNull;

public class CampaignInfo {

    public CampaignInfo() {
        this(new ClientInfo());
    }

    public CampaignInfo(ClientInfo clientInfo) {
        this.clientInfo = clientInfo;
    }

    public CampaignInfo(ClientInfo clientInfo, Campaign campaign) {
        this.clientInfo = clientInfo;
        this.campaign = campaign;
    }

    private ClientInfo clientInfo;
    private Campaign campaign;

    public ClientInfo getClientInfo() {
        return clientInfo;
    }

    public void setClientInfo(ClientInfo clientInfo) {
        this.clientInfo = clientInfo;
    }

    public CampaignInfo withClientInfo(ClientInfo clientInfo) {
        setClientInfo(clientInfo);
        return this;
    }

    public Campaign getCampaign() {
        return campaign;
    }

    public void setCampaign(Campaign campaign) {
        this.campaign = campaign;
    }

    public CampaignInfo withCampaign(Campaign campaign) {
        setCampaign(campaign);
        return this;
    }

    public Long getCampaignId() {
        return ifNotNull(campaign, Campaign::getId);
    }

    public Long getOrderId() {
        return ifNotNull(campaign, Campaign::getOrderId);
    }

    public Long getUid() {
        return getClientInfo().getUid();
    }

    public ClientId getClientId() {
        return getClientInfo().getClientId();
    }

    public Integer getShard() {
        return getClientInfo().getShard();
    }
}

File: direct/test/java/ru/yandex/direct/core/testing/info/campaign/CampaignInfo.kt
package ru.yandex.direct.core.testing.info.campaign

import ru.yandex.direct.core.entity.campaign.model.CommonCampaign
import ru.yandex.direct.core.testing.info.CampaignInfo
import ru.yandex.direct.core.testing.info.CampaignInfoConverter
import ru.yandex.direct.core.testing.info.ClientInfo
import ru.yandex.direct.core.testing.info.UserInfo

abstract class CampaignInfo<T : CommonCampaign>(
    var managerInfo: UserInfo?,
    var agencyInfo: UserInfo?,
    clientInfo: ClientInfo,
    var typedCampaign: T
) : CampaignInfo(clientInfo) {
    val shard: Int get() = clientInfo.shard
    val uid: Long get() = clientInfo.uid

    //val clientId: ClientId? get() = clientInfo.clientId
    val campaignId: Long get() = typedCampaign.id
    val id: Long get() = campaignId

    open fun withTypedCampaign(typedCampaign: T) = apply { this.typedCampaign = typedCampaign }

    fun toCampaignInfo(): CampaignInfo {
        return CampaignInfoConverter.toCampaignInfo(clientInfo, typedCampaign)
    }
}
